File: test_shopping.js
const { JSDOM, VirtualConsole } = require("jsdom");
const fs = require("fs");
const path = require("path");

const html = fs.readFileSync(path.join(__dirname, "shopping_list.html"), "utf8");

const PASS = "\x1b[32m✓\x1b[0m";
const FAIL = "\x1b[31m✗\x1b[0m";
const results = [];

function check(name, cond, detail = "") {
  const icon = cond ? PASS : FAIL;
  console.log(`  ${icon} ${name}` + (detail ? `  (${detail})` : ""));
  results.push([name, cond]);
}

function makeDom(preload = null) {
  const dom = new JSDOM(html, {
    runScripts: "dangerously",
    resources: "usable",
    url: "http://localhost/shopping_list.html",
    beforeParse(window) {
      if (preload) window.localStorage.setItem("shopping-list-v1", JSON.stringify(preload));
    },
  });
  return dom;
}

// 헬퍼: 입력 → 추가
function addItem(doc, text) {
  const inp = doc.getElementById("new-item");
  const btn = doc.getElementById("add-btn");
  inp.value = text;
  btn.dispatchEvent(new doc.defaultView.MouseEvent("click", { bubbles: true }));
}

function addItemEnter(doc, text) {
  const inp = doc.getElementById("new-item");
  inp.value = text;
  inp.dispatchEvent(
    new doc.defaultView.KeyboardEvent("keydown", { key: "Enter", bubbles: true })
  );
}

// ────────────────────────────────────────────────────────────────────────────
// 1. 초기 상태
// ────────────────────────────────────────────────────────────────────────────
console.log("\n[1] 초기 상태");
{
  const { document: doc } = makeDom().window;
  check("빈 상태 메시지 표시", !!doc.querySelector(".empty"));
  check("헤더 요약 '항목 없음'", doc.getElementById("summary").textContent.includes("항목 없음"));
  check("아이템 0개", doc.querySelectorAll(".item").length === 0);
}

// ────────────────────────────────────────────────────────────────────────────
// 2. 아이템 추가
// ────────────────────────────────────────────────────────────────────────────
console.log("\n[2] 아이템 추가");
{
  const { document: doc } = makeDom().window;

  addItem(doc, "사과");
  const labels = () => doc.querySelectorAll(".item label");
  check("버튼 클릭으로 항목 추가", labels().length === 1);
  check("추가된 텍스트 확인", labels()[0].textContent === "사과");

  addItemEnter(doc, "바나나");
  check("Enter 키로 항목 추가", labels().length === 2);

  addItem(doc, "우유");
  check("3개 항목 추가", labels().length === 3);

  check("헤더 요약 '전체 3개'", doc.getElementById("summary").textContent.includes("전체 3개"));

  // 공백 무시
  addItem(doc, "   ");
  check("공백 입력 무시", labels().length === 3);

  // 추가 후 입력창 비워짐
  addItem(doc, "딸기");
  check("추가 후 입력창 비워짐", doc.getElementById("new-item").value === "");
}

// ────────────────────────────────────────────────────────────────────────────
// 3. 체크(완료) 기능
// ────────────────────────────────────────────────────────────────────────────
console.log("\n[3] 체크(완료) 기능");
{
  const dom2 = makeDom();
  const doc2 = dom2.window.document;

  addItem(doc2, "사과");
  addItem(doc2, "바나나");
  addItem(doc2, "우유");

  const items = () => doc2.querySelectorAll(".item");
  const cbs   = () => doc2.querySelectorAll(".item input[type=checkbox]");

  // 체크
  cbs()[0].checked = true;
  cbs()[0].dispatchEvent(new dom2.window.Event("change", { bubbles: true }));
  check("체크 후 done 클래스 추가", items()[0].classList.contains("done"));
  check("체크박스 checked 상태", cbs()[0].checked);

  const doneText = doc2.getElementById("done-count").textContent;
  check("완료 카운트 '1 / 3 완료'", doneText.includes("1 / 3 완료"), doneText);

  // 체크 해제
  cbs()[0].checked = false;
  cbs()[0].dispatchEvent(new dom2.window.Event("change", { bubbles: true }));
  check("체크 해제 후 done 클래스 제거", !items()[0].classList.contains("done"));
}

// ────────────────────────────────────────────────────────────────────────────
// 4. 개별 삭제
// ────────────────────────────────────────────────────────────────────────────
console.log("\n[4] 개별 삭제");
{
  const dom = makeDom();
  const doc = dom.window.document;

  addItem(doc, "사과");
  addItem(doc, "바나나");
  addItem(doc, "우유");

  const delBtns = () => doc.querySelectorAll(".del-btn");
  const itemLabels = () => [...doc.querySelectorAll(".item label")].map(l => l.textContent);

  delBtns()[0].click();
  check("첫 번째 항목 삭제 후 2개 남음", doc.querySelectorAll(".item").length === 2);
  const remaining = itemLabels();
  check("나머지 항목 순서 유지 (바나나, 우유)", JSON.stringify(remaining) === JSON.stringify(["바나나", "우유"]), JSON.stringify(remaining));

  // 나머지 전부 삭제
  delBtns()[0].click();
  delBtns()[0].click();
  check("전부 삭제 후 빈 메시지 표시", !!doc.querySelector(".empty"));
}

// ────────────────────────────────────────────────────────────────────────────
// 5. 완료 항목 일괄 삭제
// ────────────────────────────────────────────────────────────────────────────
console.log("\n[5] 완료 항목 일괄 삭제");
{
  const dom = makeDom();
  const doc = dom.window.document;

  addItem(doc, "사과");
  addItem(doc, "바나나");
  addItem(doc, "우유");

  const cbs = () => doc.querySelectorAll(".item input[type=checkbox]");
  cbs()[0].checked = true;
  cbs()[0].dispatchEvent(new dom.window.Event("change", { bubbles: true }));
  cbs()[1].checked = true;
  cbs()[1].dispatchEvent(new dom.window.Event("change", { bubbles: true }));

  doc.getElementById("clear-done").click();

  const items = doc.querySelectorAll(".item");
  check("완료 2개 삭제 후 1개 남음", items.length === 1);
  check("미완료 항목(우유) 유지", items[0].querySelector("label").textContent === "우유");
}

// ────────────────────────────────────────────────────────────────────────────
// 6. localStorage 영속성
// ────────────────────────────────────────────────────────────────────────────
console.log("\n[6] localStorage 영속성");
{
  const dom = makeDom();
  const doc = dom.window.document;

  addItem(doc, "사과");
  addItem(doc, "바나나");

  const cbs = () => doc.querySelectorAll(".item input[type=checkbox]");
  cbs()[0].checked = true;
  cbs()[0].dispatchEvent(new dom.window.Event("change", { bubbles: true }));

  // localStorage에 저장된 데이터 직접 읽기
  const stored = JSON.parse(dom.window.localStorage.getItem("shopping-list-v1") || "[]");
  check("localStorage에 2개 항목 저장", stored.length === 2, `저장된 개수: ${stored.length}`);
  check("첫 번째 항목 done=true 저장", stored[0].done === true, `done: ${stored[0].done}`);
  check("두 번째 항목 done=false 저장", stored[1].done === false, `done: ${stored[1].done}`);

  // 동일 localStorage로 새 DOM 생성해 영속성 검증
  const dom2 = makeDom(stored);
  const doc2 = dom2.window.document;
  const items2 = doc2.querySelectorAll(".item");
  check("새 DOM 로드 후 항목 유지 (2개)", items2.length === 2, `실제: ${items2.length}`);
  check("체크 상태 유지 (done 클래스)", items2[0].classList.contains("done"));
}

// ────────────────────────────────────────────────────────────────────────────
// 결과 요약
// ────────────────────────────────────────────────────────────────────────────
const passed = results.filter(([, ok]) => ok).length;
const failed = results.filter(([, ok]) => !ok).length;
console.log("\n" + "─".repeat(45));
process.stdout.write(`결과: ${passed}/${results.length} 통과`);
if (failed) {
  console.log(`  |  실패: ${failed}개`);
  results.filter(([, ok]) => !ok).forEach(([name]) => console.log(`    ${FAIL} ${name}`));
} else {
  console.log("  — 모든 테스트 통과!");
}
console.log();
process.exit(failed > 0 ? 1 : 0);

File: test_shopping.py
import sys
from pathlib import Path
from playwright.sync_api import sync_playwright, expect

FILE_URL = f"file://{Path(__file__).parent / 'index.html'}"

PASS = "\033[32m✓\033[0m"
FAIL = "\033[31m✗\033[0m"

results = []

def check(name, cond, detail=""):
    icon = PASS if cond else FAIL
    print(f"  {icon} {name}" + (f"  ({detail})" if detail else ""))
    results.append((name, cond))


def run():
    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        ctx = browser.new_context()

        def fresh_page():
            page = ctx.new_page()
            page.goto(FILE_URL)
            page.evaluate("localStorage.clear()")
            page.reload()
            return page

        print("\n[1] 초기 상태")
        page = fresh_page()
        empty = page.locator(".empty")
        check("빈 상태 메시지 표시", empty.is_visible())
        check("헤더 요약 '항목 없음'", "항목 없음" in page.locator("#summary").text_content())
        page.close()

        print("\n[2] 아이템 추가")
        page = fresh_page()
        inp = page.locator("#new-item")
        add_btn = page.locator("#add-btn")

        inp.fill("사과")
        add_btn.click()
        item_labels = page.locator(".item label")
        check("버튼 클릭으로 항목 추가", item_labels.count() == 1)
        check("추가된 텍스트 확인", item_labels.first.text_content() == "사과")

        inp.fill("바나나")
        inp.press("Enter")
        check("Enter 키로 항목 추가", item_labels.count() == 2)

        inp.fill("우유")
        add_btn.click()
        check("3개 항목 추가", item_labels.count() == 3)

        check("헤더 요약 '전체 3개'", "전체 3개" in page.locator("#summary").text_content())

        inp.fill("   ")
        add_btn.click()
        check("공백 입력 무시", item_labels.count() == 3)

        inp.fill("딸기")
        add_btn.click()
        check("추가 후 입력창 비워짐", inp.input_value() == "")

        page.close()

        print("\n[3] 체크(완료) 기능")
        page = fresh_page()
        for text in ["사과", "바나나", "우유"]:
            page.locator("#new-item").fill(text)
            page.locator("#add-btn").click()

        first_cb = page.locator(".item input[type=checkbox]").first
        first_item = page.locator(".item").first

        first_cb.check()
        check("체크 후 done 클래스 추가", "done" in first_item.get_attribute("class"))
        check("체크박스 checked 상태", first_cb.is_checked())

        first_cb.uncheck()
        check("체크 해제 후 done 클래스 제거", "done" not in first_item.get_attribute("class"))

        first_cb.check()
        done_text = page.locator("#done-count").text_content()
        check("완료 카운트 표시 (1 / 3 완료)", "1 / 3 완료" in done_text, done_text)
        page.close()

        print("\n[4] 개별 삭제")
        page = fresh_page()
        for text in ["사과", "바나나", "우유"]:
            page.locator("#new-item").fill(text)
            page.locator("#add-btn").click()

        del_btns = page.locator(".del-btn")
        del_btns.first.click()
        items = page.locator(".item")
        check("첫 번째 항목 삭제 후 2개 남음", items.count() == 2)
        labels = [items.nth(i).locator("label").text_content() for i in range(2)]
        check("나머지 항목 순서 유지", labels == ["바나나", "우유"], str(labels))

        page.locator(".del-btn").first.click()
        page.locator(".del-btn").first.click()
        check("전부 삭제 후 빈 메시지 표시", page.locator(".empty").is_visible())
        page.close()

        print("\n[5] 완료 항목 일괄 삭제")
        page = fresh_page()
        for text in ["사과", "바나나", "우유"]:
            page.locator("#new-item").fill(text)
            page.locator("#add-btn").click()

        checkboxes = page.locator(".item input[type=checkbox]")
        checkboxes.nth(0).check()
        checkboxes.nth(1).check()
        page.locator("#clear-done").click()

        items = page.locator(".item")
        check("완료 2개 삭제 후 1개 남음", items.count() == 1)
        check("미완료 항목(우유) 유지", items.first.locator("label").text_content() == "우유")
        page.close()

        print("\n[6] localStorage 영속성")
        page = ctx.new_page()
        page.goto(FILE_URL)
        page.evaluate("localStorage.clear()")
        page.reload()

        for text in ["사과", "바나나"]:
            page.locator("#new-item").fill(text)
            page.locator("#add-btn").click()
        page.locator(".item input[type=checkbox]").first.check()

        page2 = ctx.new_page()
        page2.goto(FILE_URL)
        items2 = page2.locator(".item")
        check("새로고침 후 항목 유지 (2개)", items2.count() == 2)
        check("체크 상태 유지", "done" in items2.first.get_attribute("class"))
        page.close()
        page2.close()

        browser.close()

    passed = sum(1 for _, ok in results if ok)
    failed = sum(1 for _, ok in results if not ok)
    total  = len(results)
    print(f"\n{'─'*45}")
    print(f"결과: {passed}/{total} 통과", end="")
    if failed:
        print(f"  |  실패: {failed}개")
        for name, ok in results:
            if not ok:
                print(f"    {FAIL} {name}")
    else:
        print("  — 모든 테스트 통과!")
    print()
    return failed == 0


if __name__ == "__main__":
    ok = run()
    sys.exit(0 if ok else 1)